File: ccswitch.py
#!/usr/bin/env python3
"""
ccswitch.py — CC Switch(v2, com.ccswitch.desktop)一键注册。

不直接读写 CC Switch 的 ~/.cc-switch/cc-switch.db(SQLite),
而是生成其官方支持的 deeplink `ccswitch://v1/import?...`,
交由运行中的 CC Switch 自行落库,对 schema 变化免疫。
格式与本机已有实现一致(spark_api / sub2api 的 ccswitchImport.ts):

    resource=provider&app=claude&name=…&homepage=…&endpoint=…
    &apiKey=…[&model=…]&configFormat=json&usageEnabled=false
"""

from __future__ import annotations

import platform
import subprocess
import sys
import urllib.parse


def build_deeplink(*,
                   endpoint: str,
                   name: str = "WorkBuddy 算力网关",
                   api_key: str = "workbuddy",
                   model: str | None = None,
                   app: str = "claude") -> str:
    """构造 CC Switch 导入 deeplink。endpoint 不含 /v1/messages(由协议拼接)。"""
    params: list[tuple[str, str]] = [
        ("resource", "provider"),
        ("app", app),
        ("name", name),
        ("homepage", endpoint),
        ("endpoint", endpoint),
        ("apiKey", api_key),
    ]
    if model:
        params.insert(2, ("model", model))
    params += [
        ("configFormat", "json"),
        ("usageEnabled", "false"),
    ]
    return "ccswitch://v1/import?" + urllib.parse.urlencode(params)


def open_deeplink(url: str) -> bool:
    """把 deeplink 交给本机:macOS 用 open、Linux 用 xdg-open、Windows 用 start。"""
    if platform.system() == "Darwin":
        cmd = ["open", url]
    elif sys.platform.startswith("linux"):
        cmd = ["xdg-open", url]
    elif platform.system() == "Windows":
        cmd = ["cmd", "/c", "start", "", url]
    else:
        return False
    try:
        subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        return True
    except Exception:
        return False


if __name__ == "__main__":
    print(build_deeplink(endpoint="http://127.0.0.1:8787", model="hy3", api_key="workbuddy"))

File: src-tauri/src/ccswitch.rs
/// CC Switch deeplink — mirrors ccswitch.py
use urlencoding::encode;

pub fn build_deeplink(
    endpoint: &str,
    name: &str,
    api_key: &str,
    model: Option<&str>,
) -> String {
    let mut params: Vec<(String, String)> = vec![
        ("resource".into(), "provider".into()),
        ("app".into(), "claude".into()),
        ("name".into(), name.into()),
        ("homepage".into(), endpoint.into()),
        ("endpoint".into(), endpoint.into()),
        ("apiKey".into(), api_key.into()),
    ];
    if let Some(m) = model {
        if !m.is_empty() {
            params.insert(2, ("model".into(), m.into()));
        }
    }
    params.push(("configFormat".into(), "json".into()));
    params.push(("usageEnabled".into(), "false".into()));

    let qs: String = params
        .iter()
        .map(|(k, v)| format!("{}={}", encode(k), encode(v)))
        .collect::<Vec<_>>()
        .join("&");
    format!("ccswitch://v1/import?{qs}")
}

pub fn open_deeplink(url: &str) -> bool {
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg(url)
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .spawn()
            .is_ok()
    }
    #[cfg(target_os = "linux")]
    {
        std::process::Command::new("xdg-open")
            .arg(url)
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .spawn()
            .is_ok()
    }
    #[cfg(target_os = "windows")]
    {
        std::process::Command::new("cmd")
            .args(["/c", "start", "", url])
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .spawn()
            .is_ok()
    }
    #[cfg(not(any(target_os = "macos", target_os = "linux", target_os = "windows")))]
    {
        let _ = url;
        false
    }
}
